platforms/common/uORB/uORBDeviceNode.hpp

```diff
@@ -243,6 +240,24 @@ class DeviceNode
 
 	const char *get_devname() const {return _devname;}
 
+	int8_t cb_dequeue(uorb_cb_handle_t cb)
+	{
+		IndexedStackHandle<CB_LIST_T> callbacks(_callbacks);
+		EventWaitItem *item = callbacks.peek(cb);
+#ifndef CONFIG_BUILD_FLAT
+		int8_t ret = -1;
+
+		if (item->cb_triggered) {
+			item->cb_triggered = false;
```

Comment:
Possible race here ? As boolean is non-atomic, it can/will take multiple instructions / cycles to do the read/modify/write operation. Thus, can devicenode be writing true to this at the same time as cb_dequeue is setting it to false, i.e. is the node lock taken prior to cb_dequeue or no ?

Using an atomic type (like uint64_t or uintptr_t) would ensure there is no race here, as the rmw operation is guaranteed atomic

Reply:
I don't think that the race matters here (but I can be wrong!):

1) The "item" is subscriber specific structure. There is only one reader thread (the subsciber) and 1..N writer threads (but typically just one).
2) the if can only evaluate to true or false. I don't see an intermediate state where this could go wrong...
3) the race between reader and writer(s) occurs only when device node "write" turns the boolean from false->true in the middle of the above lines. 
4) a race between multiple writers will end up with cb_triggered = true
5) in a race between multiple writers and a reader there is a chance of the "if" evaluating to true because of another writer has just set it. And then a second writer is just setting it in the middle and that change would get lost when the reader sets it to false. But again, this is not a problem because the subscriber already got notified that there are updates (the if already evaluated to true).

Please let me know what other races there might be if any!

uint64_t is not an atomic type on 32-bit targets... uintptr_t on the other hand is a bit messy for this purpose.

One thinkable way would be to use e.g. unsigned and have a counter like this:
 if (__atomic_load_n(&item->cb_triggered, __ATOMIC_SEQ_CST) > 0) {
                       __atomic_fetch_sub(&item->cb_triggered, 1, __ATOMIC_SEQ_CST);

(Or _atomic_store). But it looks a bit... iffy.

Please let me know how the boolean can fail?

Reply:
If the non-atomic rmw operation does not matter then it is fine!

Reply:
My logic above is correct in the sense that in the callback, the subscriber gets info about all the updates. However, if the subscriber is written in a way that it expects to get a new call fro each publish (e.g. popping one data from uORB queue in each call, this would be wrong.

I will change the boolean to atomic counter after all, then it works the same as in flat build, where there is really one call for each publish